x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/support/CartesianShapeValuesSourceType.java

```diff
@@ -0,0 +1,114 @@
+import org.elasticsearch.script.AggregationScript;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.FieldContext;
+import org.elasticsearch.search.aggregations.support.MissingValues;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+import org.elasticsearch.xpack.spatial.index.fielddata.CartesianShapeValues;
+import org.elasticsearch.xpack.spatial.index.fielddata.IndexCartesianPointFieldData;
+import org.elasticsearch.xpack.spatial.index.fielddata.IndexShapeFieldData;
+import org.elasticsearch.xpack.spatial.index.fielddata.ShapeValues;
+import org.elasticsearch.xpack.spatial.index.fielddata.plain.CartesianShapeIndexFieldData;
+
+import java.io.IOException;
+
+public class CartesianShapeValuesSourceType extends ShapeValuesSourceType {
+
+    static CartesianShapeValuesSourceType INSTANCE = new CartesianShapeValuesSourceType();
+
+    public static CartesianShapeValuesSourceType instance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public ValuesSource getEmpty() {
+        return CartesianShapeValuesSource.EMPTY;
+    }
+
+    @Override
+    public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script, AggregationContext context) {
+        boolean isPoint = fieldContext.indexFieldData() instanceof IndexCartesianPointFieldData;
+        boolean isShape = fieldContext.indexFieldData() instanceof IndexShapeFieldData;
+        if (isPoint == false && isShape == false) {
+            throw new IllegalArgumentException(
+                "Expected point or shape type on field ["
+                    + fieldContext.field()
+                    + "], but got ["
+                    + fieldContext.fieldType().typeName()
+                    + "]"
+            );
+        }
+        if (isPoint) {
+            return new CartesianPointValuesSource.Fielddata((IndexCartesianPointFieldData) fieldContext.indexFieldData());
+        }
+        return new CartesianShapeValuesSource.Fielddata((CartesianShapeIndexFieldData) fieldContext.indexFieldData());
+    }
+
+    @Override
+    public ValuesSource replaceMissing(
+        ValuesSource valuesSource,
+        Object rawMissing,
+        DocValueFormat docValueFormat,
+        AggregationContext context
+    ) {
+        // TODO: This code is almost entirely in common between Geo and Cartesian
+        CartesianShapeValuesSource shapeValuesSource = (CartesianShapeValuesSource) valuesSource;
+        final ShapeValues.ShapeValue missing = CartesianShapeValues.EMPTY.missing(rawMissing.toString());
+        return new CartesianShapeValuesSource() {
+            @Override
+            public CartesianShapeValues shapeValues(LeafReaderContext context) {
+                ShapeValues values = shapeValuesSource.shapeValues(context);
```

Comment:
I would really like that CartesianShapeValuesSource returns a CartesianShapeValues. There are methods in ShapeValues that are geo specific and should only be present in GeoShapeValues. There is something that feels not right at the moment.

Reply:
The previous PR was supposed to make sure that ShapeValues had no geo-specific code, but I can already see that I left at least `CoordinateEncoder.Geo` there, so I'll review this and fix it. I suspect there will be more cases where the previous PR did not quite complete the generalization of the code.

Reply:
I notice also the use of `LatLonGeometry` and `LatLonGeometryRelationVisitor`. However, despite their names, those two classes are not geo-specific and look like they work perfectly well for cartesian data too. Perhaps they should be renamed?

Reply:
The biggest problem is not in ShapeValues but ShapeValue. For example the method # relate(LatLonGeometry latLonGeometry) is for geo. 

I think GeoshapeValue and Cartisianshapevalue are meant to support different things and should not be represented with the same ShapeValue.

Reply:
But that is what I meant above. The method `relate(LatLonGeometry)` does not have anything geo-specific in it. All that code, and the code it calls is specific to the Lucene `Component2D` mechanism and is identical between geo and cartesian. The only mistake here is that the classes are incorrectly named `LatLon...`, while they have nothing to do with geo, and are used for both geo and cartesian data. They all work directly on the `Geometry` and `Component2D` interfaces, both of which are used for both geo and cartesian data.

So I think I can _fix_ this by renaming the classes prefixed with `LatLon` something else. Then it will be more clear.

Reply:
I looked at this again and discovered that the generalization I did in the previous PR actually completed the necessary generalization. There is no need for instances of `CartesianShapeValues` or `GeoShapeValues` at all. The geo and cartesian behaviour had already been extracted and provided as parameters directly to the `ShapeValues` constructor. This meant I could simply remove those classes, although I kept them as factory classes with static methods. And I could move most of the common code directly into `ShapeValues`. Three blocks of duplicated code have been moved reducing code duplication quite a lot:

* Empty shape values
* Shape values based on BinaryDocValues
* Shape values based on wrapping other ShapeValues and replacing missing values

This improvement is pushed in https://github.com/elastic/elasticsearch/pull/89216/commits/678c3472602b3092a86399f4c828390d24ec65e7

Reply:
I have still not fixed the incorrectly named `LatLon*` classes that have nothing to do with Geo, and are used by both geo and cartesian. That could be done anytime. It is tech debt that precedes this work.

Reply:
I am getting confused, I think you are not working on top of the latest main, still exposing ShapeValues on geo / cartesian classes

Reply:
Yes, indeed. But I've fixed that now. Reverted the commit that moved back to GeoShapeValues.

Reply:
That commit was based on the assumption that ShapeValues has geo-specific code in it. And that is still not true. I remember when I approved that before I said that I wanted to re-evaluate my assumptions and reconsider that work once we did the Cartesian PR. Now this is the Cartesian PR and it looks to me like the generalization is in fact a very good idea. We make a whole stack of code common between Cartesian and Geo. And this is all enabled by the Lucene Component2D interface which is what actually makes this stack common.

Reply:
I think you assumptions are wrong, please don't expose ShapeValues as being the same for geo and cartesian, they are not.